subst/react-native.js.flow

```diff
@@ -37,6 +40,44 @@ export type {
   SyntheticEvent as NativeSyntheticEvent,
 } from 'react-native/Libraries/Types/CoreEventTypes';
 
+// In real RN, this takes type params, so you can give a list of valid event
+// names it should know about:
+//
+//   class EmitterSubscription<EventDefinitions: {...}, K: $Keys<EventDefinitions>>
+//     extends _EventSubscription<EventDefinitions, K>
+//     implements EventSubscription {
+//
+// Not so in @types/react-native:
+//
+//   interface EmitterSubscription extends EventSubscription {
+//
+// It just uses `string` for an event name. So, match that unspecificity by
+// choosing unspecific type params.
```

Comment:
> it looks like `EmitterSubscription` of a more-specific set of event types shouldn't be usable as an `EmitterSubscription` of a less-specific set of event types, nor vice versa.

This makes sense, and it's convincing. I'm really glad for this opportunity to think about variance without my head hurting! 😅

I may have had thoughts on this type that I don't remember right now, but I'm pretty sure it came up when I tried to translate [`@react-native-clipboard/clipboard`](https://www.npmjs.com/package/@react-native-clipboard/clipboard) with TsFlower. Its code does have a [module](https://github.com/react-native-clipboard/clipboard/blob/master/src/NativeClipboard.ts) where `EmitterSubscription` appears in its exports.

I'm not sure yet what kind of code would try to use an `EmitterSubscription` of a more-specific set of event types as an `EmitterSubscription` of a less-specific set of event types, or vice versa. Do you have a case in mind, either a priori or after looking at that clipboard library? Or have I misunderstood something, and you expect this definition of `EmitterSubscription` to be impossible to use at all?

Reply:
> However: I don't yet know what kind of code might try to use an `EmitterSubscription` of a more-specific set of event types as an `EmitterSubscription` of a less-specific set of event types, or vice versa. Do you have a case in mind, either a priori or after looking at that clipboard library?

Yeah -- I mean that this definition in the PR:
```js
declare export class EmitterSubscription
  extends _EmitterSubscription<{| +[string]: mixed |}, string> {}
```
defines a TS-facing `EmitterSubscription` as being an actual-RN `EmitterSubscription` with a maximally-broad set of event types. Which is what the comment says about the motivation for it:
> So, match that unspecificity by choosing unspecific type params.

So what I'm expecting is that if some library provides a thing that would be a valid RN `EmitterSubscription` with some particular more-specific set of event types, then that will not be usable as an `EmitterSubscription` according to this definition.

Reply:
I agree that we'd expect a Flow error in code that tries to use an `EmitterSubscription<{| +foo: … |}>`, for example, as an `EmitterSubscription` according to this definition. I just don't yet see it being likely for real code to actually do something like that.

Reply:
So I think it makes sense to try this as an experiment.

Now I'm wondering if my code comment is quite right, though, hmm.

> If these types work in the context where you try to use them, then perhaps I'm just missing something here. If they don't, then I think the fallback will be to use `any` in these type parameters. After all, that faithfully reflects what these `@types/react-native` definitions themselves say; for example:
> 
> ```js
> interface EmitterSubscription extends EventSubscription {
>     emitter: EventEmitter;
>     listener: () => any;
> ```

Here's the full definition in `@types/react-native`:

<details>
  <summary>(Collapsed)</summary>

```js
/**
 * EmitterSubscription represents a subscription with listener and context data.
 */
interface EmitterSubscription extends EventSubscription {
    emitter: EventEmitter;
    listener: () => any;
    context: any;

    /**
     * @param emitter - The event emitter that registered this
     *   subscription
     * @param subscriber - The subscriber that controls
     *   this subscription
     * @param listener - Function to invoke when the specified event is
     *   emitted
     * @param context - Optional context object to use when invoking the
     *   listener
     */
    new (
        emitter: EventEmitter,
        subscriber: EventSubscriptionVendor,
        listener: () => any,
        context: any,
    ): EmitterSubscription;

    /**
     * Removes this subscription from the emitter that registered it.
     * Note: we're overriding the `remove()` method of EventSubscription here
     * but deliberately not calling `super.remove()` as the responsibility
     * for removing the subscription lies with the EventEmitter.
     */
    remove(): void;
}
```

</details>

And RN's implementation, in `react-native/Libraries/vendor/emitter/_EmitterSubscription.js`:

<details>
  <summary>(Collapsed)</summary>

```js
/**
 * EmitterSubscription represents a subscription with listener and context data.
 */
class EmitterSubscription<EventDefinitions: {...}, K: $Keys<EventDefinitions>>
  extends _EventSubscription<EventDefinitions, K>
  implements EventSubscription {
  emitter: EventEmitter<EventDefinitions>;
  listener: ?(...$ElementType<EventDefinitions, K>) => mixed;
  context: ?$FlowFixMe;

  /**
   * @param {EventEmitter} emitter - The event emitter that registered this
   *   subscription
   * @param {EventSubscriptionVendor} subscriber - The subscriber that controls
   *   this subscription
   * @param {function} listener - Function to invoke when the specified event is
   *   emitted
   * @param {*} context - Optional context object to use when invoking the
   *   listener
   */
  constructor(
    emitter: EventEmitter<EventDefinitions>,
    subscriber: EventSubscriptionVendor<EventDefinitions>,
    listener: (...$ElementType<EventDefinitions, K>) => mixed,
    context: ?$FlowFixMe,
  ) {
    super(subscriber);
    this.emitter = emitter;
    this.listener = listener;
    this.context = context;
  }

  /**
   * Removes this subscription from the emitter that registered it.
   * Note: we're overriding the `remove()` method of _EventSubscription here
   * but deliberately not calling `super.remove()` as the responsibility
   * for removing the subscription lies with the EventEmitter.
   */
  remove(): void {
    this.emitter.__removeSubscription(this);
  }
}
```

</details>

It's odd to me that the TypeScript types the listener function as `() => any`. Shouldn't there be a param for the event that gets passed to the listener? Anyway, I guess this is a place where my code comment is wrong:

```js
// [@types/react-native] just uses `string` for an event name.
```

because `@types/react-native`…just…doesn't bother with events or event names for the listener function. 🤷‍♂️

In at least one place, I think your observation that `@types/react-native` uses `any` for an event name isn't quite right. They type the `emitter` property as `EventEmitter`, and `EventEmitter` is full of `eventType: string`, not `eventType: any`, see:


<details>
  <summary>(Collapsed)</summary>

```js
declare class EventEmitter {
    /**
     *
     * @param subscriber - Optional subscriber instance
     *   to use. If omitted, a new subscriber will be created for the emitter.
     */
    constructor(subscriber?: EventSubscriptionVendor | null);

    /**
     * Adds a listener to be invoked when events of the specified type are
     * emitted. An optional calling context may be provided. The data arguments
     * emitted will be passed to the listener function.
     *
     * @param eventType - Name of the event to listen to
     * @param listener - Function to invoke when the specified event is
     *   emitted
     * @param context - Optional context object to use when invoking the
     *   listener
     */
    addListener(eventType: string, listener: (...args: any[]) => any, context?: any): EmitterSubscription;

    /**
     * Removes all of the registered listeners, including those registered as
     * listener maps.
     *
     * @param eventType - Optional name of the event whose registered
     *   listeners to remove
     */
    removeAllListeners(eventType?: string): void;

    /**
     * Removes a specific subscription. Called by the `remove()` method of the
     * subscription itself to ensure any necessary cleanup is performed.
     * @deprecated Use `remove` on the EventSubscription from `addListener`.
     */
    removeSubscription(subscription: EmitterSubscription): void;

    /**
     * Returns the number of listeners that are currently registered for the given
     * event.
     *
     * @param eventType - Name of the event to query
     */
    listenerCount(eventType: string): number;

    /**
     * Emits an event of the given type with the given data. All handlers of that
     * particular type will be notified.
     *
     * @param eventType - Name of the event to emit
     * @param Arbitrary arguments to be passed to each registered listener
     *
     * @example
     *   emitter.addListener('someEvent', function(message) {
     *     console.log(message);
     *   });
     *
     *   emitter.emit('someEvent', 'abc'); // logs 'abc'
     */
    emit(eventType: string, ...params: any[]): void;

    /**
     * Removes the given listener for event of specific type.
     *
     * @param eventType - Name of the event to emit
     * @param listener - Function to invoke when the specified event is
     *   emitted
     *
     * @example
     *   emitter.removeListener('someEvent', function(message) {
     *     console.log(message);
     *   }); // removes the listener if already registered
     * @deprecated Use `remove` on the EventSubscription from `addListener`.
     */
    removeListener(eventType: string, listener: (...args: any[]) => any): void;
}
```

</details>